File: logic/round.cpp
#include "round.h"

Round::Round(players firstPlayer) : roundDeck(), p1Points(0),p2Points(0), p1Hand(), p2Hand(), m_startCard(roundDeck.draw()),m_firstPlayer(firstPlayer)
{
    
}

int Round::getP1Points()
{
    return p1Points;
}

int Round::getP2Points()
{
    return p2Points;
}

void Round::addToP1Pile(Card cardToAdd)
{
    p1Pile.push_back(cardToAdd);
}

void Round::addToP2Pile(Card cardToAdd)
{
    p2Pile.push_back(cardToAdd);
}

void Round::countPiles()    
{
    int diamonds = 0;
    int sevens = 0;
    int sixes = 0;
    int cards = 0;
   
    cards = p2Pile.size();
    for (int i = 0; i < cards; ++i)
    {
        if (p1Pile[i].getSuit() == Card::D)
        {
            ++diamonds;
            if (p1Pile[i].getRank() == 7) //only for 7 diamond
            {
                ++p1Points;
            }
            else
            {
                ++p2Points;
            }
        }
        if (p1Pile[i].getRank() == 7)
        {
            ++sevens;
        }
        else if (p1Pile[i].getRank() == 6)
        {
            ++sixes;
        }
    }

    if (sevens > 2 || (sevens == 2 && sixes > 2))
    {
        ++p1Points;
    }
    else if (sevens < 2 || (sevens == 2 && sixes < 2))
    {
        ++p2Points;
    }
    if (cards > 20)
    {
        ++p1Points;
    }
    else if (cards < 20)
    {
        ++p2Points;
    }
    if (diamonds > 5)
    {
        ++p1Points;
    }
    else if (diamonds < 5)
    {
        ++p2Points;
    }

}

void Round::firstMiniRound(bool choice)  //aka first mini-round
{
    if (choice == false)
    {
        m_board.addToBoard(m_startCard);
        for (int i = 0; i < NUM_OF_HAND; ++i)
        {
            p1Hand.addToHand(roundDeck.draw());
            p2Hand.addToHand(roundDeck.draw());
            m_board.addToBoard(roundDeck.draw());
        }
    }
    else
    {
        if (m_firstPlayer == P1)
        {
            p1Hand.addToHand(m_startCard);
            p2Hand.addToHand(roundDeck.draw());
        }
        else
        {
            p2Hand.addToHand(m_startCard);
            p1Hand.addToHand(roundDeck.draw());
        }
        for (int i = 0; i < NUM_OF_HAND-1; ++i)
        {
            p1Hand.addToHand(roundDeck.draw());
            p2Hand.addToHand(roundDeck.draw());
        }
        for (int i = 0; i < NUM_OF_BOARD; ++i)
        {
            m_board.addToBoard(roundDeck.draw());
        }

    }
    
}

void Round::giveCardsToPlayers()
{
    for (int i = 0; i < NUM_OF_HAND; ++i)
    {
        p1Hand.addToHand(roundDeck.draw());
        p2Hand.addToHand(roundDeck.draw());
    }


}

File: app/src/main/java/com/dinari/shkuba/Round.kt
package com.dinari.shkuba

class Round(firstPlayer: Int) {
    // Native pointer to the C++ Round instance
    private var nativeHandle: Long = 0

    init {
        nativeHandle = nativeCreate(firstPlayer)
    }

    // JNI: Create C++ Round instance
    private external fun nativeCreate(firstPlayer: Int): Long

    // JNI: Clean up C++ Round instance
    private external fun nativeDestroy(handle: Long)

    // JNI: Get P1 points
    external fun getP1Points(): Int

    // JNI: Get P2 points
    external fun getP2Points(): Int

    // JNI: Execute first mini round
    external fun firstMiniRound(choice: Boolean)

    // JNI: Give cards to players
    external fun giveCardsToPlayers()

    // JNI: Count piles and add points
    external fun countPiles()

    // JNI: Add card to P1 pile
    external fun addToP1Pile(suit: Int, rank: Int)

    // JNI: Add card to P2 pile  
    external fun addToP2Pile(suit: Int, rank: Int)

    // JNI: Get the start card
    external fun getStartCard(): NativeCard

    // Helper methods for enum support
    fun addToP1Pile(card: NativeCard) {
        addToP1Pile(card.getSuit(), card.getRank())
    }

    fun addToP2Pile(card: NativeCard) {
        addToP2Pile(card.getSuit(), card.getRank())
    }

    protected fun finalize() {
        if (nativeHandle != 0L) {
            nativeDestroy(nativeHandle)
            nativeHandle = 0L
        }
    }

    companion object {
        // Player enum values matching C++
        const val P1 = 0
        const val P2 = 1

        // Library loaded in MainActivity
    }
}